src/sql/src/pure.rs

```diff
@@ -224,74 +224,87 @@ pub async fn purify_create_source(
-                            normalize::unresolved_object_name(upstream_name.clone())?;
-                        let schemas =
-                            tables_by_name
-                                .get(&partial_upstream_name.item)
-                                .ok_or_else(|| {
-                                    sql_err!(
-                                        "table {} not found in upstream database",
-                                        partial_upstream_name
-                                    )
-                                })?;
-
-                        let (schema, table) = match &partial_upstream_name.schema {
-                            Some(schema) => match schemas.get(schema) {
-                                Some(table) => (schema, table),
-                                None => {
-                                    return Err(sql_err!(
-                                        "schema {} does not exist in upstream database",
-                                        schema
-                                    )
-                                    .into())
-                                }
+                        let schemas = match tables_by_name.get(&upstream_name.item) {
+                            Some(schemas) => schemas,
+                            None => bail!("table {upstream_name} not found in source"),
+                        };
+
+                        let schema = match &upstream_name.schema {
+                            Some(schema) => schema,
+                            None => match schemas.keys().exactly_one() {
+                                Ok(schema) => schema,
+                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the schema"),
                             },
-                            None => match schemas.iter().exactly_one() {
-                                Ok((schema, table)) => (schema, table),
-                                Err(_) => {
-                                    return Err(sql_err!(
-                                        "table {} is ambiguous, consider specifying the schema",
-                                        upstream_name
-                                    )
-                                    .into())
-                                }
+                        };
+
+                        let databases = match schemas.get(schema) {
+                            Some(databases) => databases,
+                            None => bail!("schema {schema} not found in source"),
+                        };
+
+                        let database = match &upstream_name.database {
+                            Some(database) => database,
+                            None => match databases.keys().exactly_one() {
+                                Ok(database) => database,
+                                Err(_) => bail!("table {upstream_name} is ambiguous, consider specifying the database"),
```

Comment:
It can't happen with postgres sources, but conceptually a multi-output source presents a list of objects that are named with the usual SQL rules. So it's possible to write a multi-output source that provides the following objects: `database1.foo.bar` and `database2.foo.bar`. Then if the user typed `FOR TABLES (foo.bar)` it would hit this error